### PR TITLE
[UXTHEME] Implement various Vista+ functions, mostly from WINE

### PR DESCRIPTION
**Pull request by Splitwirez**
## Purpose

Implement various uxtheme functions added in Vista or later - most are from WINE.

JIRA issue: None...? (how)

#### Minutiae
- [x] Rename `uxtheme_nt6plus.spec` to `uxtheme_vista.spec`
- [x] Determine if changes to `buffer.c` are meaningful, gate behind `DLL_EXPORT_VERSION` if so
- [x] Remove stray line (no.636) in `system.c`
- [x] Double-check for stray commented junk

#### Big important stuff
- [x] Fix NT 5.2 builds
- [x] Fix MSVC hack in `buffer.c`

#### DarkFire come pick me up im scared
- [x] Find out what (if anything) is needed by way of tests
- KVM [x] https://reactos.org/testman/compare.php?ids=102434,102441
-  KVM x64 [x] https://reactos.org/testman/compare.php?ids=102434,102441

### CODE REVIEW

File: dll/win32/uxtheme/CMakeLists.txt [hunk 1]
```diff
@@ -1,7 +1,12 @@
 
 include_directories(${REACTOS_SOURCE_DIR}/sdk/include/reactos/wine)
 add_definitions(-D__WINESRC__ -D__ROS_LONG64__)
-spec2def(uxtheme.dll uxtheme.spec ADD_IMPORTLIB)
+
+if(DLL_EXPORT_VERSION GREATER_EQUAL 0x600)
+    spec2def(uxtheme.dll uxtheme_nt6plus.spec ADD_IMPORTLIB)
```
Comment (SergeGautherie) [inline dll/win32/uxtheme/CMakeLists.txt:6]: `vista` is usually used, instead of `nt6plus`.

```suggestion
    spec2def(uxtheme.dll uxtheme_vista.spec ADD_IMPORTLIB)
```

File: dll/win32/uxtheme/buffer.c [hunk 2]
```diff
@@ -19,6 +19,28 @@
  */
 
 #include "uxthemep.h"
+#include <wine/heap.h>
```
Comment (JoachimHenze) [inline dll/win32/uxtheme/buffer.c:22]: I do think the actual code-changes should also be guarded by something like
`#if DLL_EXPORT_VERSION >= _WIN32_WINNT_VISTA`

File: dll/win32/uxtheme/system.c [hunk 3]
```diff
@@ -633,6 +633,35 @@ BOOL WINAPI IsThemeActive(void)
     return bActive;
 }
 
+
```
Comment (JoachimHenze) [inline dll/win32/uxtheme/system.c:636]:
```suggestion
```

File: dll/win32/uxtheme/buffer.c (hunk 2 above)
Comment (DarkFire01) [reply dll/win32/uxtheme/buffer.c:22]: If this was synced via the script? no.
If this was synced via copy and paste? That’s fine since they won’t be used unless that is set anyway.

@Splitwirez which did you do?

File: dll/win32/uxtheme/buffer.c [hunk 4]
```diff
@@ -41,35 +63,107 @@ HRESULT WINAPI BufferedPaintUnInit(VOID)
...
-                                       BP_PAINTPARAMS *pPaintParams,
-                                       HDC *phdc)
+HPAINTBUFFER WINAPI BeginBufferedPaint(HDC targetdc, const RECT *rect,
+        BP_BUFFERFORMAT format, BP_PAINTPARAMS *params, HDC *retdc)
 {
-    static int i;
+#if (defined(_MSC_VER))
+    RGBQUAD rgb;
+    //TODO: Fix this for MSVC
+    __debugbreak();
```
Comment (HBelusca) [inline dll/win32/uxtheme/buffer.c:72]: is this what Wine had?

File: sdk/include/psdk/uxtheme.h [hunk 5]
```diff
@@ -15,6 +15,23 @@ extern "C" {
 #define DTBG_COMPUTINGREGION        0x00000010
 #define DTBG_MIRRORDC               0x00000020
 #define DTT_GRAYED                  0x00000001
+#if DLL_EXPORT_VERSION >= _WIN32_WINNT_VISTA
+#define DTT_TEXTCOLOR    0x00000001
```
Comment (HBelusca) [inline sdk/include/psdk/uxtheme.h:19]: - Are these defines present in the official uxtheme header?
- Please keep the same whitespace indentation of the values as the other values above;
- `DLL_EXPORT_VERSION` shouldn't be used in the "public" headers. Are you looking for `_WIN32_WINNT` ?
And please put parentheses around the preprocessor test: `#if (whatever >= something)`

**Review by HBelusca — COMMENTED** [review]
Builders don't seem to be happy.

File: dll/win32/uxtheme/buffer.c (hunk 4 above)
Comment (DarkFire01) [reply dll/win32/uxtheme/buffer.c:72]: oh this is leftover from working fixing build ,
@Splitwirez  You're going to have to spend a little time finding a genuine fix to this and putting __REACTOS__ brackets around it like I do here

File: dll/win32/uxtheme/CMakeLists.txt (hunk 1 above)
Comment (Splitwirez) [reply dll/win32/uxtheme/CMakeLists.txt:6]: I feel like this naming convention is going to come back to bite us - `_vista` vs `_vistaplus` is the linguistic equivalent of `== vista` vs `>= vista`, after all.

That said...it *is* convention, so I'll change the name shortly.

Comment (cbialorucki) [reply dll/win32/uxtheme/CMakeLists.txt:6]: For the record I would also prefer to use the term "nt6" instead of "vista"
But it's not a big deal and it's the current convention used so...
¯\_(ツ)_/¯

File: dll/win32/uxtheme/buffer.c (hunk 2 above)
Comment (JoachimHenze) [reply dll/win32/uxtheme/buffer.c:22]: ???
Ofc they will be used, the implementation of e.g. BeginBufferedPaint(), ... changed, even for NT5 build.

File: dll/win32/uxtheme/CMakeLists.txt (hunk 1 above)
Comment (Splitwirez) [reply dll/win32/uxtheme/CMakeLists.txt:6]: "nt6" is still "== nt6" though...

File: dll/win32/uxtheme/buffer.c (hunk 2 above)
Comment (Splitwirez) [reply dll/win32/uxtheme/buffer.c:22]: @JoachimHenze Those aren't even exported in NT5.2, are they?

File: dll/win32/uxtheme/CMakeLists.txt (hunk 1 above)
Comment (RatinCN) [reply dll/win32/uxtheme/CMakeLists.txt:6]: It's a long story, use "_vista" is okay, that means "NT6+", not NT6 or Vista only.

File: sdk/include/psdk/uxtheme.h [hunk 6]
```diff
@@ -15,7 +15,7 @@ extern "C" {
 #define DTBG_COMPUTINGREGION        0x00000010
 #define DTBG_MIRRORDC               0x00000020
 #define DTT_GRAYED                  0x00000001
-#if DLL_EXPORT_VERSION >= _WIN32_WINNT_VISTA
+#if defined(_UXTHEME_PCH_) || (_WIN32_WINNT >= _WIN32_WINNT_VISTA)
```
Comment (HBelusca) [inline sdk/include/psdk/uxtheme.h:18]: This UXTHEME_PCH has nothing to do in public headers as well...
(I think i understand why you used it there though...)
But if you add vista+ functions in our uxtheme to be able to use them actively there, why not compiling uxtheme with its _WIN32_WINNT defined to _WIN32_WINNT_VISTA (or a higher value)?
As follows:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/shell32/CMakeLists.txt;hb=90de2b660b12919c2f60f2f31ad41e37c123d662#l9
```
remove_definitions(-D_WIN32_WINNT=0x502)
add_definitions(-D_WIN32_WINNT=0x600)
```
(Note the `remove_definitions` before in order to remove the old value.)

**Review by DarkFire01 — APPROVED** [review]
You need to fix the builders, spend some time with MSVC. But the import looks more solid now. And i can confirm it works

File: dll/win32/uxtheme/CMakeLists.txt (hunk 1 above)
Comment (Splitwirez) [reply dll/win32/uxtheme/CMakeLists.txt]: Okay, I've renamed the `uxtheme_nt6plus.spec` to `uxtheme_vista.spec`.

I still think it's a mistake to pretend people will magically somehow understand that when we use terminology that linguistically means "== Vista", we actually mean ">= Vista" (and likewise with saying "== NT 6.0" but meaning ">= NT 6.0"), but clearly that's a discussion to be had...elsewhere.

File: dll/win32/uxtheme/buffer.c (hunk 2 above)
Comment (JoachimHenze) [reply dll/win32/uxtheme/buffer.c:22]: buffer.c is compiled even for NT5.2 builds:
image
and its object file does contain BeginBufferedPaint():
image

Comment (Splitwirez) [reply dll/win32/uxtheme/buffer.c:22]: Huh...that's unexpected. I could have sworn those APIs were only introduced in Windows Vista. In fact, the official Microsoft docs even support my understanding:
- [BeginBufferedPaint](https://learn.microsoft.com/en-us/windows/win32/api/uxtheme/nf-uxtheme-beginbufferedpaint) ([archived version from Jan 7, 2022](https://web.archive.org/web/20220107031650/https://docs.microsoft.com/en-us/windows/win32/api/uxtheme/nf-uxtheme-beginbufferedpaint))
- [DrawThemeBackground](https://learn.microsoft.com/en-us/windows/win32/api/uxtheme/nf-uxtheme-drawthemebackground) ([archived version from Jan 6, 2022](https://web.archive.org/web/20220106223315/https://docs.microsoft.com/en-us/windows/win32/api/uxtheme/nf-uxtheme-drawthemebackground))
Granted, Microsoft's documentation can be very...*weird* about whether or not to acknowledge sufficiently old versions of Windows, so I'm still not quite sure whether I can trust my memory on this front.

Furthermore, they aren't exported at all at present, [even in upstream master](https://github.com/reactos/reactos/blob/master/dll/win32/uxtheme/uxtheme.spec)...unless I'm missing something here? Are those files *not* the be-all-end-all of what gets exported?

To clarify, I'm *not* trying to say that you're wrong - on the contrary...I have no idea what I'm doing, but I want to learn. I want to understand why this matters.

### In short:
Why did I think `BeginBufferedPaint` and friends don't exist in Windows Server 2003? Am I stupid?

Furthermore, if  they aren't exported in Server 2003, then...why exactly do the changes matter?

Comment (whindsaks) [reply dll/win32/uxtheme/buffer.c:22]: > I could have sworn those APIs were only introduced in Windows Vista

Buffered paint, yes but not DrawThemeBackground (it even says "Minimum supported server	Windows Server 2003 [desktop apps only]" on the current live page).

Both DrawThemeBackground and DrawThemeBackgroundEx (ordinal only?) are exported on 2003 and probably XP.

Using archive.org on pages that new is pointless, you can't use learn.microsoft.com nor docs.microsoft.com, you need to use the old msdn.microsoft.com and even then, you can't trust MSDN.

Comment (Splitwirez) [reply dll/win32/uxtheme/buffer.c:22]: @whindsaks `DrawThemeBackground` 100% exists in XP, yes. Sorry, I messed up how I phrased that - I'd only meant to mention the `DrawThemeBackground` docs to show that docs from the time period in question *do* acknowledge Server 2003 (at least sometimes?), but my head wasn't screwed on quite right so I forgot to actually say that lmao

The archive.org versions are included in case Microsoft decides to start weirdly specifically erasing history even further - in short, those links are there for tomorrow, not for today.

Comment (JoachimHenze) [reply dll/win32/uxtheme/buffer.c:22]: From my pov what everybody here wrote doesn't contradict anything.
What I write here I did not double-check, but from my understanding
- DrawThemeBackground and DrawThemeBackgroundEx (ordinal only?) are exported on 2003 and probably XP.
- but BeginBufferedPaint is not from the MS tools

What I actually want is: If you change implementation of e.g. BeginBufferedPaint(), you are free to do that, but ideally those changes should be guarded by some ifdef, so we will bake them only into the binary when the target is defined to be NT6+.
A 5.2 build of our uxtheme should ideally not have them, neither implemented nor exported.
If that exclusion isn't the case yet, then it should be added.

They were inside the binaries already before you touched the implementation of BeginBufferedPaint(), and yes, that is and was an existing flaw most likely. But since they are now changed and are beginning to be more than just stubs, the exclusion should be added now.

Comment (whindsaks) [reply dll/win32/uxtheme/buffer.c:22]: Using BeginBufferedPaint inside uxtheme is harmless IMHO. We have a time-machine of sorts here, we know BeginBufferedPaint is coming but we don't know exactly when MS started using it internally, we only know it was exported and documented starting with Vista.

Comment (JoachimHenze) [reply dll/win32/uxtheme/buffer.c:22]: This **could** be correct @whindsaks , I cannot prove your claim to be wrong. It is possible that MS used it internally already before they started exporting it. I don't know how we could verify or falsify that hypothesis though, therefore my brain defaults to `try-to-be-careful-what-we-do-add`-mode.

Comment (whindsaks) [reply dll/win32/uxtheme/buffer.c:22]: Yes but the point is, if uxtheme internally needs similar functionality (even just for [bitmap handling without drawing](https://devblogs.microsoft.com/oldnewthing/20110520-00/?p=10613)), it is better to model it after an API we know is coming.

Comment (JoachimHenze) [reply dll/win32/uxtheme/buffer.c:22]: I do agree with that part, but we should take extra-care to not break anything regarding backwards compatibility and should not bloat configurations where specific stuff might not be needed (yet). But sure, **in the long term** moving into the same direction as MS did is definitely better than cooking our own soup entirely. That doesn't invalidate the intent of my discussion-thread though.

File: dll/win32/uxtheme/draw.c [hunk 7]
```diff
@@ -1473,6 +1489,52 @@ HRESULT WINAPI DrawThemeText(HTHEME hTheme, HDC hdc, int iPartId, int iStateId,
...
+    _In_ int iCharCount,
+    _In_ DWORD dwTextFlags,
+    _Inout_ LPRECT pRect,
+    _In_ const DTTOPTS *options
+)
+{
+    TRACE("(%p %p %d %d %s:%d 0x%08lx %p %p)\n", hTheme, hdc, iPartId, iStateId,
+        debugstr_wn(pszText, iCharCount), iCharCount, dwTextFlags, pRect, options);
+    return UXTHEME_DrawThemeTextEx(hTheme, hdc, iPartId, iStateId, pszText, iCharCount, dwTextFlags, pRect, options);
+}
```
Comment (binarymaster) [inline dll/win32/uxtheme/draw.c:1536]: Hmm... why there's a wrapper DrawThemeTextEx => UXTHEME_DrawThemeTextEx, it doesn't seem to be required.

File: sdk/include/psdk/uxtheme.h [hunk 8]
```diff
@@ -90,13 +126,57 @@ typedef struct _MARGINS {
...
+    _In_ int iStateId,
+    _In_ LPCWSTR pszText,
+    _In_ int iCharCount,
+    _In_ DWORD dwTextFlags,
+    _Inout_ LPRECT pRect,
+    _In_ const DTTOPTS *options
+);
+
+
+
```
Comment (binarymaster) [inline sdk/include/psdk/uxtheme.h:179]: For the consistency with surrounding definitions.
```suggestion
/* Vista+ */
HRESULT WINAPI DrawThemeTextEx(HTHEME,HDC,int,int,LPCWSTR,int,DWORD,LPRECT,const DTTOPTS*);
/* Vista- */
```

**Comment by binarymaster** [comment]
> JIRA issue: None...? (how)

I guess I'm particularly interested in getting DrawThemeTextEx function in, because it will allow drawing "Text in window" in the DrawNCPreview properly - that is [CORE-5991](https://jira.reactos.org/browse/CORE-5991).

Windows uses the "Button" part to draw this text with a shadow, but overrides the text color by using DrawThemeTextEx internally with some flags (it's not exported in Windows XP/2003), if the selected theme defines it - that's the case for Lautus, Lunar, and Modern.

File: sdk/include/psdk/uxtheme.h (hunk 8 above)
Comment (HBelusca) [reply sdk/include/psdk/uxtheme.h:179]: Usually @binarymaster we tend to sanitize the new stuff we put in the public headers, and make them wine-ugly.

Comment (binarymaster) [reply sdk/include/psdk/uxtheme.h:179]: Alright then.

**Comment by binarymaster** [comment]
DrawThemeTextEx has been merged in https://github.com/reactos/reactos/commit/d4dc0cb54596f99b8fc40c90232823801b363cff and utilized in the next commit.

Resolved conflict here with a merge commit.

**Comment by Splitwirez** [comment]
> This PR is stale because it received no updates in the last 4 months. Without removing the stale label, or commenting on this ticket it will be closed in 2 weeks.

BUMP??? PLEASE NO??????? I TOTALLY FORGOT I HADN'T ALREADY WRAPPED THIS UP @DarkFire01 I AM SO SORRY SGIPJSFGOIHSDJHODFIJHODJI

File: dll/win32/uxtheme/system.c [hunk 9]
```diff
@@ -888,6 +930,23 @@ HRESULT WINAPI SetWindowTheme(HWND hwnd, LPCWSTR pszSubAppName,
...
 
+#if (DLL_EXPORT_VERSION >= _WIN32_WINNT_VISTA)
+/***********************************************************************
+ *      SetWindowThemeAttribute                             (UXTHEME.@)
+ */
+HRESULT
+WINAPI
+SetWindowThemeAttribute(
+    _In_ HWND hwnd,
+    _In_ enum WINDOWTHEMEATTRIBUTETYPE type,
```
Comment (HBelusca) [inline dll/win32/uxtheme/system.c:941]: eAttribute

File: dll/win32/uxtheme/buffer.c [hunk 10]
```diff
@@ -91,38 +183,65 @@ HRESULT WINAPI BufferedPaintSetAlpha(HPAINTBUFFER hBufferedPaint, const RECT *pr
...
+    TRACE("(%p)\n", buffer);
+
+    return buffer ? buffer->memorydc : NULL;
 }
 
 /***********************************************************************
  *      GetBufferedPaintTargetDC                           (UXTHEME.@)
  */
-HDC WINAPI GetBufferedPaintTargetDC(HPAINTBUFFER hBufferedPaint)
+HDC WINAPI GetBufferedPaintTargetDC(HPAINTBUFFER bufferhandle)
```
Comment (HBelusca) [inline dll/win32/uxtheme/buffer.c:219]: why the change

File: dll/win32/uxtheme/buffer.c [hunk 11]
```diff
@@ -41,35 +64,104 @@ HRESULT WINAPI BufferedPaintUnInit(VOID)
 /***********************************************************************
  *      BeginBufferedPaint                                 (UXTHEME.@)
  */
-HPAINTBUFFER WINAPI BeginBufferedPaint(HDC hdcTarget,
-                                       const RECT * prcTarget,
-                                       BP_BUFFERFORMAT dwFormat,
-                                       BP_PAINTPARAMS *pPaintParams,
-                                       HDC *phdc)
+HPAINTBUFFER WINAPI BeginBufferedPaint(HDC targetdc, const RECT *rect,
+        BP_BUFFERFORMAT format, BP_PAINTPARAMS *params, HDC *retdc)
```
Comment (HBelusca) [inline dll/win32/uxtheme/buffer.c:68]: F*ck wine

File: dll/win32/uxtheme/buffer.c [hunk 12]
```diff
@@ -19,6 +19,29 @@
...
+static void free_paintbuffer(struct paintbuffer *buffer)
+{
+    DeleteObject(buffer->bitmap);
+    DeleteDC(buffer->memorydc);
+    heap_free(buffer);
+}
+
+static struct paintbuffer *get_buffer_obj(HPAINTBUFFER handle)
+{
+    return handle;
```
Comment (HBelusca) [inline dll/win32/uxtheme/buffer.c:43]: lol they need a full blown function just to do a cast?

File: dll/win32/uxtheme/uxtheme_vista.spec [hunk 13]
```diff
@@ -0,0 +1,128 @@
+1 stdcall -noname QueryThemeServices()
```
[the remark this answers is not in the document]
Comment (DarkFire01) [reply dll/win32/uxtheme/uxtheme_vista.spec:1]: The ordinals matter because software devs started doing STUPID stuff with uxtheme after vista.

[the remark this answers is not in the document]
Comment (Splitwirez) [reply dll/win32/uxtheme/uxtheme_vista.spec:1]: this is so sad can we get conditional ordinals or something like that in `.spec` files


Okay, but seriously...is that possible? Or if not, is it *at least* possible to use more than one `.spec` file simultaneously, so we could have e.g. a `uxtheme_common.spec` and then a separate one for stuff that varies between Windows releases?

[the remark this answers is not in the document]
Comment (whindsaks) [reply dll/win32/uxtheme/uxtheme_vista.spec:1]: Or handle the ordinals at run-time by checking the process target Windows version and jump to the correct function.

File: dll/win32/uxtheme/buffer.c (hunk 11 above)
Comment (Splitwirez) [reply dll/win32/uxtheme/buffer.c:68]: I get that you're not a fan of the way WINE does things in some cases...but this PR is a for ReactOS, and I *already* have no idea what I'm doing, even without the added confusion.

Were you trying to help me or help get this PR wrapped up and merged by sharing your distaste for WINE in a comment here? If so, then...how?

Were you implying a recommended course of action on my part?

File: dll/win32/uxtheme/buffer.c (hunk 12 above)
Comment (Splitwirez) [reply dll/win32/uxtheme/buffer.c:43]: ...is this supposed to be actionable feedback for this PR?

File: dll/win32/uxtheme/buffer.c [hunk 14]
```diff
@@ -91,38 +183,65 @@ HRESULT WINAPI BufferedPaintSetAlpha(HPAINTBUFFER hBufferedPaint, const RECT *pr
...
+    TRACE("(%p)\n", buffer);
+
+    return buffer ? buffer->targetdc : NULL;
 }
 
 /***********************************************************************
  *      GetBufferedPaintTargetRect                         (UXTHEME.@)
  */
-HRESULT WINAPI GetBufferedPaintTargetRect(HPAINTBUFFER hBufferedPaint, RECT *prc)
+HRESULT WINAPI GetBufferedPaintTargetRect(HPAINTBUFFER bufferhandle, RECT *rect)
```
[the remark this answers is not in the document]
Comment (SergeGautherie) [reply dll/win32/uxtheme/buffer.c:231]: `dll/win32/uxtheme             # Forked`.
Meanwhile, if code is copied from WINE, having it same as upstream should be preferred indeed.

File: sdk/include/psdk/uxtheme.h [hunk 15]
```diff
@@ -57,6 +56,8 @@ extern "C" {
 
 typedef HANDLE HPAINTBUFFER;
 typedef HANDLE HTHEME;
+//Vista+
```
Comment (HBelusca) [inline sdk/include/psdk/uxtheme.h]:
```suggestion
// Vista+
```
Maybe add a _win32_winnt and/or a dll_export_ ... compile check?

File: sdk/include/psdk/uxtheme.h [hunk 16]
```diff
@@ -15,8 +15,7 @@ extern "C" {
 #define DTBG_COMPUTINGREGION        0x00000010
 #define DTBG_MIRRORDC               0x00000020
 #define DTT_GRAYED                  0x00000001
-
-/* DTTOPTS.dwFlags bits */
+//Vista+
```
Comment (HBelusca) [inline sdk/include/psdk/uxtheme.h]:
```suggestion
// Vista+
```
And same comment as above.

File: sdk/include/psdk/uxtheme.h [hunk 17]
```diff
@@ -109,6 +125,7 @@ typedef struct _MARGINS {
     int cyBottomHeight;
 } MARGINS, *PMARGINS;
 
+//Vista+
```
Comment (HBelusca) [inline sdk/include/psdk/uxtheme.h]:
```suggestion
// Vista+
```
ditto

File: sdk/include/psdk/uxtheme.h [hunk 18]
```diff
@@ -126,6 +143,13 @@ typedef struct _DTTOPTS {
     DTT_CALLBACK_PROC pfnDrawTextCallback;
     LPARAM lParam;
 } DTTOPTS, *PDTTOPTS;
+typedef struct _BP_ANIMATIONPARAMS
```
Comment (HBelusca) [inline sdk/include/psdk/uxtheme.h:148]:
```suggestion

typedef struct _BP_ANIMATIONPARAMS
```

File: sdk/include/psdk/uxtheme.h [hunk 19]
```diff
@@ -192,6 +216,25 @@ DrawThemeTextEx(
     _Inout_ LPRECT pRect,
     _In_ const DTTOPTS *options
 );
+
+// vista+
```
Comment (HBelusca) [inline sdk/include/psdk/uxtheme.h]:
```suggestion
// Vista+
```
with compile time version check?

**Review by HBelusca — APPROVED** [review]
Globally ok, here are some suggestions for adding compile time version checks , and there are some remaining PR todos that need to be handled apparently.

File: sdk/include/psdk/uxtheme.h (hunk 19 above)
Comment (DarkFire01) [reply sdk/include/psdk/uxtheme.h]: No.

File: dll/win32/uxtheme/uxtheme_vista.spec (hunk 13 above)
[the remark this answers is not in the document]
Comment (HBelusca) [reply dll/win32/uxtheme/uxtheme_vista.spec:1]: If the ordinals change, it might be possible to do something like this:
```
1 stdcall -version=<some_range> foo()
1 stdcall -version=<other_range> bar()
```

File: dll/win32/uxtheme/system.c [hunk 20]
```diff
@@ -633,6 +633,34 @@ BOOL WINAPI IsThemeActive(void)
...
+{
+    BOOL bIsCompositionActive;
+    DWMISCOMPOSITIONENABLED pDwmIsCompositionEnabled;
+    HANDLE hdwmapi = GetModuleHandleW(L"dwmapi.dll");
+
+    if (!hdwmapi)
+    {
+        hdwmapi = LoadLibraryW(L"dwmapi.dll");
+        if (!hdwmapi)
+            ERR("Failed to load dwmapi\n");
```
[the remark this answers is not in the document]
Comment (HBelusca) [reply dll/win32/uxtheme/system.c:667]: @darkfire any feedback on this?

[the remark this answers is not in the document]
Comment (DarkFire01) [reply dll/win32/uxtheme/system.c:667]: good point, this was resolved

File: dll/win32/uxtheme/buffer.c [hunk 21]
```diff
@@ -91,38 +183,65 @@ HRESULT WINAPI BufferedPaintSetAlpha(HPAINTBUFFER hBufferedPaint, const RECT *pr
...
+    *width = buffer->rect.right - buffer->rect.left;
+
+    return S_OK;
 }
 
 /***********************************************************************
  *      GetBufferedPaintDC                                 (UXTHEME.@)
  */
-HDC WINAPI GetBufferedPaintDC(HPAINTBUFFER hBufferedPaint)
+HDC WINAPI GetBufferedPaintDC(HPAINTBUFFER bufferhandle)
```
[the remark this answers is not in the document]
Comment (DarkFire01) [reply dll/win32/uxtheme/buffer.c:206]: It's in wine like this, ill change these if we ever rewrite our own code.

File: sdk/include/psdk/uxtheme.h (hunk 15 above)
Comment (DarkFire01) [reply sdk/include/psdk/uxtheme.h]: This has been talked about off and on
For sure I'm not going to add active checks for vista RTM-SP2 header changes anymore when they don't effect anything.

I think // Vista+ is fair. but honestly If your goal is to retain the ability to compile an NT5.2 OS these changes to the sdk won't effect that :)

**Review by DarkFire01 — APPROVED** [review]
I barely made any changes really, 
WINE'isms sucks sometimes but its not our job to push that on PR authors to change. 
Good work, sorry it took so long.

**Review by cbialorucki — APPROVED** [review]

File: dll/win32/uxtheme/uxtheme_vista.spec (hunk 13 above)
[the remark this answers is not in the document]
Comment (whindsaks) [reply dll/win32/uxtheme/uxtheme_vista.spec:1]: This would imply uxtheme_8, uxtheme_10, uxtheme_11 etc?

I don't know if there are any collisions but I know the dark theme stuff has been a moving target.

[the remark this answers is not in the document]
Comment (DarkFire01) [reply dll/win32/uxtheme/uxtheme_vista.spec:1]: No, 
Actually this is a whole conversation that needs to happen.
We have a WIP patch to enable vista+ theme format loading but its all a mess and moving target like you said. For now I'd just like to see us get past the vista hump, and solve this issue another day.
